### ISSUE TITLE
Representation of photon emitter button in PDOM

### ISSUE THREAD

**Issue by jessegreenberg** [phetsims/molecules-and-light#296]
From #295 and mentioned over slack, how would we like this button to appear in the PDOM? @terracoda is going to provide some markup to try and then I will listen to how each sounds on our supported platforms.

**Comment by terracoda** [comment]
@jessegreenberg, I read this article a while back, and I think we could explore some of these ideas for the different kinds of buttons we use in sims...
https://www.smashingmagazine.com/2017/09/building-inclusive-toggle-buttons/

I'll share the article and then post in some html code tomorrow.

**Comment by terracoda** [comment]
There are 3 or 4 types of toggle buttons in that article :-)
Main point: communicate a change in state on one way only - state change or name change.

When a label does not change` aria-pressed` is often a good choice.
However, when a label change is needed (e.g. "On/Off" or "Pause/Play") changing the state and label at the same time is not a good choice.

**Comment by terracoda** [comment]
@jessegreenberg I think we might have three options for the Light Source button. The first 2 do not require any visual changes. The third would require having on/off on the screen.
**Options**
1. Traditional toggle button with aria-pressed=”true/false” - a traditional toggle button is an acceptable option because the current button interaction does not have a visual name change and the on/off states of the button are communicated visually  with a pressed/unpressed look. 
2. On/Off switch - button with aria-pressed=”true/false” and aria-role=”switch”, the pressed state should be communicated as on and off.
3. On/Off switch - button with aria-role=”switch” and aria-labeledby and coded labels for on and off.

**Code Option 1 **
Note that the dynamic part of the lable on the button is not about the state change. `aria-pressed="true"` or `aria-pressed="false"` communicate the change in state.
```html
<button type="button" aria-pressed="true">  
  {{Infrared}} Light Source
</button> 
```
**Code Option 2 **
Use a switch role on button to communicate on and off instead of pressed and not pressed. I am not sure what screen readers will support this option, and not sure if `aria-checked="true"` could work here as well.
```html
<button type="button" aria-pressed="true" role="switch">  
  {{Infrared}} Light Source
</button> 
```
**Code Option 3 **
```html
<div>
<span id="light-source-name">{{Infrared}} Light Source</span>
    <button role="switch" aria-checked="true" aria-labelledby="light-source-name">
      <span>on</span>
      <span>off</span>
    </button>
</div> 
```

**Comment by terracoda** [comment]
@jessegreenberg, I am curious to know if any of these options work well in MAL.

**Comment by terracoda** [comment]
We also need new help text. Here are the 5 options that are in the design doc:

1. Turn light source on or off.
2. Turn light source photons on or off.
3. Turn light source on to start photons.
4. Activate light source to start photons.
5. Activate light source to start exploration.

Just go with option 1 for now.

**Comment by terracoda** [comment]
@jessegreenberg, is it easy to prototype options 1 and 2? I would think option 3 is more involved.

**Comment by emily-phet** [comment]
@terracoda - is there a reason to *not* go with option 1? Seems like it would be good to keep this as simple as possible, and #1 seems the simplest (from my naive interpretation).

**Comment by terracoda** [comment]
@emily-phet, I think option 1 will work and it is consistant with the visual design of the button. If we want to literally communicate "off" and "on", option 2 might be an option.

It's not an issue yet, but I am not sure how options 1 or 2 will work with voice activation software.

Option 3, might be the clearer way to communicate "off" and "on" and would work with voice activation software.

**Comment by terracoda** [comment]
We have many different types of buttons in PhET sims and that article I posted above has a few design patterns that could potentially be very useful for us.

**Comment by terracoda** [comment]
@jessegreenberg, could you prototype options 1 and 2 first, please.

**Comment by terracoda** [comment]
And we need a label on the button as soon as possible, too. It is currently nameless and thus inaccessible. 

The dynamic name for the button is similar the dynamic name for the slider, so hopefully some of that code is re-usable for the new button interaction.

New dynamic label for button from design doc:
- {{Microwave}} Light Source
- {{Infrared}} Light Source
- {{Visible}} Light Source
- {{Ultraviolet}} Light Source

**Comment by terracoda** [comment]
Please re-assign if you have questions :-)

**Comment by terracoda** [comment]
@jessegreenberg, @Matthew-Moore240 liked option 5 for the button help text. Let's use that for now.

**Comment by terracoda** [comment]
@jessegreenberg, the help text is changed to option 3:
- Turn light source on to start photons.

**Comment by jessegreenberg** [comment]
OK, done in the above commits. I added
1) an accessible name for the button
2) The aria-pressed attribute to the sticky button
3) Help text (5) from https://github.com/phetsims/molecules-and-light/issues/296#issuecomment-598240042
4) Optionally the `switch` role to the button, which can be added with query parameter ?switch.

Back to you to review these two options @terracoda. I tested on NVDA with Firefox and the output is identical with and without ?switch.

**Comment by terracoda** [comment]
Review of button with Voice Over
**Option 1 - regular toggle button**
- `aria-pressed="false"` read out as toggle button on focus
  - Infrared Light Source, toggle button
- Upon activation, a change to `aria-pressed="true"` - no announcement of state change
- `aria-pressed="true"` - on refocus "selected" is read out confirming the state change. 
  - Infrared Light Source, selected, toggle button

That's not as good as I expected! I thought the state change would be announced automatically.

**Option 2 - toggle button with role="switch"**
- Sounds really nice on focus:
  -  Infrared Light Source, off, switch

@jessegreenberg , upon refocus the state is not changed, It still says "off, switch". I think the aria-pressed attribute is not changing - it still said "false" in the code.

Can you have a look?

In both cases, I don't think the changed state is automatically announced - at least not by Voice Over. I think it will be beneficial to implement the UI context responses for this button regardless of which option we eventually go with.

I'll start a new issue for that which you can work on once option 2 is re-checked.

Also, it would be good to make note in this issue what the utterances are for JAWS and NVDA for these two options. Would you might doing that once you have checked/fixed Option 2?

**Comment by jessegreenberg** [comment]
I verified that with ?switch the aria-pressed state is changing.
When not pressed:
![image](https://user-images.githubusercontent.com/6396244/77551616-f5670b00-6e88-11ea-96b7-61aecc8ebdbd.png)
When pressed:
![image](https://user-images.githubusercontent.com/6396244/77551693-06178100-6e89-11ea-893e-3368bc278364.png)

**Comment by jessegreenberg** [comment]
Results with NVDA + Firefox:
Option 1:
- (Focus When not pressed): Infrared Light Source, toggle button, not pressed
  - (Pressing without moving focus): "Pressed"
- (Focus when pressed): Infrared Light Source, toggle button, pressed
  - (Pressing without moving focus): "Not Pressed"

Option 2: Identical to option 1.

**Comment by jessegreenberg** [comment]
Over slack, @terracoda and I discovered that we are supposed to use aria-checked with role=switch rather than aria-pressed. Re-testing Option 2:
**Results with NVDA + Firefox:**
Option 2:
- (Focus When not pressed): Infrared Light Source, toggle button, not pressed
  - (Pressing without moving focus): "Checked"
- (Focus when pressed): Infrared Light Source, toggle button, not pressed, checked
  - (Pressing without moving focus): "Not checked"

**Results with JAWS + Chrome**
Option 2:
- (Focus When not pressed): Infrared Light Source, switch off
  - (Pressing without moving focus): "Pressed on"
- (Focus when pressed): "Infrared LIght Source, switch, pressed on"
  - (Pressing without moving focus): "Off"

**Comment by jessegreenberg** [comment]
Trying to improve NVDA, we tried using both aria-pressed AND aria-checked. State is reported correctly, but "pressed" and "checked" are both included.
**Results with NVDA + Firefox:**
Option 2:
- (Focus When not pressed): Infrared Light Source, toggle button, not pressed
  - (Pressing without moving focus): "Pressed, checked"
- (Focus when pressed): "Infrared Light Source, toggle button, pressed, checked
  - (Pressing without moving focus): "Not pressed, not checked"

**Results with JAWS + Chrome**
Option 2:
- (Focus When not pressed): Infrared Light Source, switch off
  - (Pressing without moving focus): "Pressed on"
- (Focus when pressed): "Infrared LIght Source, switch, pressed on"
  - (Pressing without moving focus): "Off"

So NVDA is now correct but a little verbose, and JAWS sounds the same as https://github.com/phetsims/molecules-and-light/issues/296#issuecomment-603909224 (quite good IMO).

**Comment by jessegreenberg** [comment]
Over slack @terracoda said that VO still sounds OK with both aria-pressed and aria-checked attributes. Unfortunately, the state change response is lost when aria-pressed is included. But the role and state are still conveyed correctly on focus.

We are going to proceed with 
```html
<button type="button" aria-pressed="true" aria-checked="true" role="switch">  
  {{Infrared}} Light Source
</button> 
```

**Comment by terracoda** [comment]
Just confirming and documenting that with Safari and VO, the states and switch role are communicated nicely on focus:
- Infrared Light Source, off, switch
- Infrared Light Source, on, switch

Without the extra attribute, VO also communicates the state change upon activation:
- on, Infrared Light Source
- off, Infrared Light Source

Once the bug in NVDA is fixed, we will be able to remove the unnecessary `aria-pressed` attribute.

Until then, we could supplement the interaction with UI context responses like:
- Photons on.
- Photons off.

I will create a separate issue for the UI context responses.
Sound good, @jessegreenberg?

**Comment by terracoda** [comment]
Just noting this other issue in a11y-research repo as a reminder of the discussion that have already taken place about buttons that communicate their state changes in different ways https://github.com/phetsims/a11y-research/issues/137#issuecomment-482796376

**Comment by jessegreenberg** [comment]
Alright, the testing query parameter has been removed and the representation in https://github.com/phetsims/molecules-and-light/issues/296#issuecomment-603926435 is now used in master. This issue can be closed.
